### ISSUE TITLE
duplicate log level between main binary and the standard library

### ISSUE THREAD

**Issue by amtoine** [nushell/nushell#8829]
### Related problem

the original discussion appeared in https://github.com/nushell/nushell/issues/8823#issuecomment-1501032490.

### Describe the solution you'd like

we had a few ideas
- remove `NU_LOG_LEVEL` from `std`
- deprecate `--log-level` from the `nu` binary
- leave both but have `--log-level` set the default value of `NU_LOG_LEVEL

i'd vote for the third right now :yum:

**Comment by fdncred** [comment]
From my perspective, I'm not interested in deprecating --log-level as a parameter to nushell. It's too important to understanding the flow of code, especially when tracing through the ~6k line parser.rs code. 

Not to mention, as I've said in the past, it's used in performance measurement of startup when used with the debug level. This used to be more clear when using the --perf parameter, but others didn't like it, so now it's even more obscure.

**Comment by amtoine** [comment]
agree with that :+1:

**Comment by rgwood** [comment]
I think my preference would be to remove `NU_LOG_LEVEL` and use `--log-level` for both Rust logging and stdlib logging; a flag is more discoverable than an environment variable.

**Comment by amtoine** [comment]
> I think my preference would be to remove `NU_LOG_LEVEL` and use `--log-level` for both Rust logging and stdlib logging; a flag is more discoverable than an environment variable.
>
how would you adapt the log level from inside `nushell` to run `nushell` scripts?

looks to me you would have to start a new instance of `nu` :thinking:

**Comment by rgwood** [comment]
> how would you adapt the log level from inside `nushell` to run `nushell` scripts?

Hmm, depends on the details of how stdlib commands are invoked (I haven't looked into that yet, apologies). The log level could be passed to the stdlib command as a parameter or an environment variable, I think. So maybe `NU_LOG_LEVEL` would live on as an implementation detail.

**Comment by amtoine** [comment]
> The log level could be passed to the stdlib command as a parameter or an environment variable, I think. So maybe `NU_LOG_LEVEL` would live on as an implementation detail.
>
this is what i had in mind in the first place i think :ok_hand:

**Comment by bobhy** [comment]
I think we're going back and forth here because we actually have 2 separate logging functions, controlled by different flags, which log different kinds of operation: 
* --log-level for internal interpreter operation, 
*  the NU_LOG_LEVEL and `std log` for logging the operation of custom commands (eventually it will be in widespread use, not now, though).
We don't have to conflate them!  And considering the different operations they affect, maybe we shouldn't try?

For me, the only wrinkle is I do not know whether there's any logging from built in commands.  Can anybody say?  I would like to be able to say there is just one switch that controls logging for all commands, both custom and built-in, that seems like a clean division of labor.

My use case for the current `NU_LOG_LEVEL` and `std log` is this:
I want the convention to be that `std log info` is for verbose progress messages from commands.  These are suppressed by default, but can be enabled for a specific invocation.  The environment setting prefix is just the ticket for this:
```
〉def test_logging [] {
::: print "Doing something"
::: log info "verbose message about what I just did"
::: }
〉let-env NU_LOG_LEVEL = WARN    # (currently required because of #8823)
〉test_logging
Doing something
〉NU_LOG_LEVEL=INFO test_logging
Doing something
INF|2023-04-09T12:40:32.644|verbose message about what I just did
```

So, I think the right way to go is for:
1. User can set NU_LOG_LEVEL in `env.nu`
2. `nu --log-level` overrides that setting, if specified
3. `std log` tests to test the (current scope) value of NU_LOG_LEVEL.
But I want to be able to troubleshoot a custom command (e.g from std) with some kind of override. 
Imagine my custom command is sprinkled with `log info` calls, e.g:
```shell
 Currently, this works just fine (imagine the call references :
```shell
NU_LOG_LEVEL=info log info I am the walrus
```

**Comment by rgwood** [comment]
> --log-level for internal interpreter operation,

`--log-level` is currently used for any Rust logging that uses [the `log` crate](https://lib.rs/crates/log). I would not say it's just for the interpreter.

> For me, the only wrinkle is I do not know whether there's any logging from built in commands. Can anybody say?

Yes: 
![image](https://user-images.githubusercontent.com/26268125/230808047-940d1762-f7d2-46d1-9568-7fe4161f950c.png)
![image](https://user-images.githubusercontent.com/26268125/230808133-0e3f7fe6-b244-48dc-9f9c-1cc7d349c94b.png)
https://github.com/nushell/nushell/blob/d0a83fec693e50fbe38780aaec932f6b8cc8ed0a/crates/nu-system/src/linux.rs#L98-L99

**Comment by bobhy** [comment]
Other logging systems have the notion of `component` as well as `severity`.  Then the user can selectively enable logging in 2 dimensions, and that's mighty nice.

For example, component could be (root) module name: all stdlib modules would use component `std`, a user-written module could `userModule`, etc. Then the end user running the code could selectively enable logging from `std` and/or `userModule` (or both), as well as choosing different severity for each.

An environment variable to manage this might naturally be named `NU_LOGGING` and its values might be a record of `{ <component>: <severity> }`.
I like this idea and could spend some time implementing it as an enhancement of `std log`!  Any takers?

**Comment by fdncred** [comment]
@bobhy I'd support your compromise. Not sure how others feel.

**Comment by rgwood** [comment]
I think that adds more complexity to logging than we need and I would vote against.

**Comment by amtoine** [comment]
i like the idea as @fdncred but we should see the complexity of it before merging as @rgwood said :+1:

**Comment by fdncred** [comment]
One augmentation to the compromise is this.

In order to try and bring Reilly on board, it would be good to allow `nu --log-level <level>` and that should work for both the rust code and the stdlib code with the passed in log level. However, you could also specify the log level as `nu --log-level none:debug` and `none` would be for rust code and `debug` would be for stdlib.

So, the syntax actually turns out to be something like `nu --log-level <rust_log_level>:<stdlib_log_level>?` where `:<stdlib_log_level>` is optional and if it's not explicitly defined, it uses the `<rust_log_level>`. The only way you get a different log level for the stdlib is to explicity append `:<level>` to the parameter.

@rgwood Would you support doing it this way or are you still in the "it's too much complexity" camp?

**Comment by rgwood** [comment]
Apologies but I am not a fan of that approach. I do not think the Rust code and stdlib code should have separate log levels, and if they do have separate log levels I would prefer something more explicit than a `<rust_log_level>:<stdlib_log_level>` syntax (nobody will understand that at first glance).

  That said this not a hill I want to die on, I'm gonna leave this decision to Amtoine and other people who have been more involved in the stdlib than me.

**Comment by fdncred** [comment]
@rgwood no worries. Thanks for your input. Just trying to reach some type of a compromise. I know you're not just being contrary, and you want is best for nushell. Same with me. Good luck deciding @amtoine - LOL 🤣

**Comment by amtoine** [comment]
@fdncred @rgwood :scream: :laughing: 

i think i share how @rgwood feels about that :thinking: 

i do not see `std` being that much separate from the "core" `nushell` => more an implementation detail as @rgwood said iirc, especially if we're mainly moving commands from `rust` to `nushell`!


and as is say it, i think more and more that we should only have `--log-level` on the `nu` binary and have `NU_LOG_LEVEL` align on the value given when starting the shell

what follows is that the log format inside `std log` should align with the current one from the `rust` implementation :thinking: 

:bulb: that can always be changed quite easily as these are only changes from the library and not from core :wink: 

**TL;DR**: i would vote for removing the `NU_LOG_LEVEL` and use the value given by `--log-level` :yum:

**Comment by fdncred** [comment]
Good luck. Pro tip: Don't put 'trace' level entries in your scripts because you'll never find them with `nu --log-level trace`.

**Comment by bobhy** [comment]
I second @fdncred.  
We absolutely, positively have to logically independent logging settings for interpreter, standard library and user written libraries so you can enable just one and not get all the others.  Having said that, there are some logging systems that always log everything (to a file), then provide a filtering viewer, so you can find the needle in the haystack.  Why incur all that overhead all the time?

My proposal on it is described in #8853, but other solutions could be designed.

**Comment by bobhy** [comment]
Loosely, by "interpreter" I mean the parser, exec and repl.  For me the built in commands are not core to the interpreter, i think the logging for builtins should be configured independent of logging configured for "interpreter".

**Comment by amtoine** [comment]
> Good luck. Pro tip: Don't put 'trace' level entries in your scripts because you'll never find them with `nu --log-level trace`.
>
and there are a lot of `trace`s :scream: 
gotcha :relieved:

**Comment by Kissaki** [comment]
For me, the separation into `nu --log-level` and `$env.NU_LOG_LEVEL` as nushell itself vs `std/log` for anything you would add, include, and script yourself made practical sense.

Yes, `$env.NU_LOG_LEVEL` could be renamed for clarity. Maybe `$env.NU_STDLOG_LEVEL`.

#8853 suggests a full fledged logging solution. Which I guess would work too, but add implementation and configuration complexity.

While the `nu --log-level` vs `$env.NU_LOG_LEVEL` split is surprising, I assume users usually don't need or want `nu --log-level`. That makes `$env.NU_LOG_LEVEL` a very simple and obvious-usage solution which seems potentially well-suited for a shell and scripting environment.

Will Nushell move towards a more interconnected software solution that would need selective logging? Will the standard library and user solutions for frameworks grow while including logging to a degree that you would want to control their log levels separately from your own?

It seems it's at least not foreseeable as approaching right now - from my limited view.

---

The full-fledged solution suggested in #8853 may be worthwhile if the component-based log level configuration is simple and performant enough, and usage with component-specific loggers works transparently or without much effort.

Could a `std/log` transparently automatically use the current module where the log statement is as the log scope/component?

In C# with dependency injection you use `ILogger<ThisClassType>` to get a component-scoped logger. Nushell doesn't have dependency injection. Creating named loggers and putting them into variables could be cumbersome. It would be nice if it worked transparently, without additional boilerplate.

---

I guess for a `nu --log-level` to work a default configuration would have to ignore Nushell interpreter and std lib, and then the parameter only influences other/own components.
